# Demo Data

## Overview
The following table lists scSVAS applications' demo datasets. 

|scSVAS Applications | Demo Datasets| Description |
|--|--|--|
|CNV View | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|CNV Heatmap | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|Cell Phylogeny | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|Ploidy Stairstep | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|Ploidy Distribution | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|Embedding Map | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|Time Lineage | AML, HGSOC_P7 | AML: raw data from Ding *et al*. [3], input files obtained from Smith *et al*. [4] </br> HGSOC_P7: raw data from McPherson *et al*. [5], input files obtained from Smith *et al*. [4]|
|Space Lineage | HGSOC_P1,P7, PC_A21 | HGSOC_P1, P7: raw data from McPherson *et al*. [5], input files obtained from Smith *et al*. [4] </br> PC_A21: raw data from Gundem *et al*. [6], input files obtained from Smith *et al*. [4]|
|Space Prevalence | HGSOC_P1,P7, PC_A21 | HGSOC_P1, P7: raw data from McPherson *et al*. [5], input files obtained from Smith *et al*. [4] </br> PC_A21: raw data from Gundem *et al*. [6], input files obtained from Smith *et al*. [4]|
|Clonal Lineage | TNBC_T10,T16 | Raw data obtained from Navin *et al*. [1], processed by SCYN [2] and scSVAS auxiliary scripts. |
|Recurrent Event | LC | The lung cancer data present in "Recurrent Event" is prepared in another manuscript. |

Aberrations: Triple-negative breast cancer (TNBC); Acute myeloid leukaemia (AML); High grade serous ovarian cancer (HGSOC); Prostate cancer (PC); Lung cancer (LC). 

## Reference

[1] Navin, N., Kendall, J., Troge, J., Andrews, P., Rodgers, L., McIndoo, J., Cook, K., Stepansky, A., Levy, D., Esposito, D., Muthuswamy, L., Krasnitz, A., McCombie, W. R., Hicks, J., & Wigler, M. (2011). Tumour evolution inferred by single-cell sequencing. *Nature*, 472(7341), 90–94. https://doi.org/10.1038/nature09807

[2] Feng, X., Chen, L., Qing, Y., Li, R., Li, C., & Li, S. C. (2020). SCYN: Single cell CNV profiling method using dynamic programming. *bioRxiv*. https://doi.org/10.1101/2020.03.27.011353

[3] Ding, L., Ley, T. J., Larson, D. E., Miller, C. A., Koboldt, D. C., Welch, J. S., Ritchey, J. K., Young, M. A., Lamprecht, T., McLellan, M. D., McMichael, J. F., Wallis, J. W., Lu, C., Shen, D., Harris, C. C., Dooling, D. J., Fulton, R. S., Fulton, L. L., Chen, K., Schmidt, H., … DiPersio, J. F. (2012). Clonal evolution in relapsed acute myeloid leukaemia revealed by whole-genome sequencing. *Nature*, 481(7382), 506–510. https://doi.org/10.1038/nature10738

[4] Smith, M. A., Nielsen, C. B., Chan, F. C., McPherson, A., Roth, A., Farahani, H., Machev, D., Steif, A., & Shah, S. P. (2017). E-scape: interactive visualization of single-cell phylogenetics and cancer evolution. *Nature methods*, 14(6), 549–550. https://doi.org/10.1038/nmeth.4303

[5] McPherson, A., Roth, A., Laks, E., Masud, T., Bashashati, A., Zhang, A. W., Ha, G., Biele, J., Yap, D., Wan, A., Prentice, L. M., Khattra, J., Smith, M. A., Nielsen, C. B., Mullaly, S. C., Kalloger, S., Karnezis, A., Shumansky, K., Siu, C., Rosner, J., … Shah, S. P. (2016). Divergent modes of clonal spread and intraperitoneal mixing in high-grade serous ovarian cancer. *Nature genetics*, 48(7), 758–767. https://doi.org/10.1038/ng.3573

[6] Gundem, G., Van Loo, P., Kremeyer, B., Alexandrov, L. B., Tubio, J., Papaemmanuil, E., Brewer, D. S., Kallio, H., Högnäs, G., Annala, M., Kivinummi, K., Goody, V., Latimer, C., O'Meara, S., Dawson, K. J., Isaacs, W., Emmert-Buck, M. R., Nykter, M., Foster, C., Kote-Jarai, Z., … Bova, G. S. (2015). The evolutionary history of lethal metastatic prostate cancer. *Nature*, 520(7547), 353–357. https://doi.org/10.1038/nature14347

